### PR TITLE
Refactor GetBalanceReportInfo to use callback for result

### PR DESCRIPTION
**Pull request by jdkuki**
Resolves brave/brave-browser#5142

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/brave-ios/Ledger/BATBraveLedger.h [hunk 1]
```diff
@@ -188,8 +188,9 @@ NS_SWIFT_NAME(BraveLedger)
 
 @property (nonatomic, readonly) NSDictionary<NSString *, BATBalanceReportInfo *> *balanceReports;
 
-- (BATBalanceReportInfo *)balanceReportForMonth:(BATActivityMonth)month
-                                           year:(int)year;
+- (void)balanceReportForMonth:(BATActivityMonth)month
+                              year:(int)year
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.h]: Minor nit: Obj-C formatting, colon's should be aligned on new lines (Xcode would do this, Visual Studio Code probably doesnt know how)

File: vendor/brave-ios/Ledger/BATBraveLedger.mm [hunk 2]
```diff
@@ -617,11 +617,12 @@ - (void)onGrantFinish:(ledger::Result)result grant:(ledger::GrantPtr)grant
   return bridgedReports;
 }
 
-- (BATBalanceReportInfo *)balanceReportForMonth:(BATActivityMonth)month year:(int)year
+- (void)balanceReportForMonth:(BATActivityMonth)month year:(int)year completion:(void (NS_NOESCAPE ^)(bool result, BATBalanceReportInfo * _Nullable info))completion
```
Comment (kylehickinson) [inline vendor/brave-ios/Ledger/BATBraveLedger.mm]: Let's just drop the `bool result` in the completion block. We'll know whether or not it failed based on if `info` is nil or not.

Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: As it is, info will never be nil per `bat_publishers.cc`: 
```
void BatPublishers::getBalanceReport(
    ledger::ACTIVITY_MONTH month,
    int year,
    ledger::GetBalanceReportCallback callback) {
  ledger::BalanceReportInfo info;
  bool result =
    getBalanceReportInternal(month, year, &info);
  callback(result, info.Clone());
}
```

Are you suggesting it should be: 
```
void BatPublishers::getBalanceReport(
    ledger::ACTIVITY_MONTH month,
    int year,
    ledger::GetBalanceReportCallback callback) {
  ledger::BalanceReportInfo info;
  bool result =
    getBalanceReportInternal(month, year, &info);
  callback(result ? info.Clone() : nullptr);
}
```
?

File: vendor/brave-ios/Ledger/BATBraveLedger.h (hunk 1 above)
Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.h]: fixed

File: vendor/brave-ios/Ledger/BATBraveLedger.mm (hunk 2 above)
Comment (kylehickinson) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: I didn't necessarily mean the underlying C++ had to change, just that for in Obj-C we use nullability checks a lot, so its simpler. (In Obj-C calling or accessing `nil` doesn't result in a crash).

So just meant that for the Obj-C completion block can just be called like `completion(result ? bridgedInfo : nil)`

Comment (jdkuki) [reply vendor/brave-ios/Ledger/BATBraveLedger.mm]: gotcha, fixed

**Pull request merged**